File: packages/api/src/test/lib/replicate-database.worker.test.ts
import type { ITaskContext } from "task-queue";
import type { IReplicateDatabaseData, IReplicateProgressMessage } from "../../lib/replicate-database.types";

// ── module mocks ─────────────────────────────────────────────────────────────

jest.mock("../../lib/open-storage", () => ({
    openStorage: jest.fn(),
}));

jest.mock("../../lib/media-file-database", () => ({
    createMediaFileDatabase: jest.fn().mockReturnValue({
        bsonDatabase: { __label: "source-bson" },
        metadataCollection: {},
    }),
}));

jest.mock("../../lib/replicate", () => ({
    replicate: jest.fn().mockResolvedValue({
        filesImported: 0,
        copiedFiles: 0,
        copiedRecords: 0,
        prunedFiles: [],
    }),
}));

jest.mock("utils", () => ({
    log: { verbose: jest.fn(), error: jest.fn(), exception: jest.fn(), info: jest.fn() },
}));

import { openStorage } from "../../lib/open-storage";
import { replicate } from "../../lib/replicate";
import { replicateDatabaseHandler } from "../../lib/replicate-database.worker";

const mockOpenStorage = openStorage as jest.MockedFunction<typeof openStorage>;
const mockReplicate = replicate as jest.MockedFunction<typeof replicate>;

//
// Builds a minimal ITaskContext for testing.
//
function makeContext(overrides: Partial<ITaskContext> = {}): ITaskContext {
    return {
        uuidGenerator: { generate: jest.fn().mockReturnValue("test-uuid") },
        timestampProvider: { now: jest.fn().mockReturnValue(Date.now()), dateNow: jest.fn().mockReturnValue(new Date()) },
        sessionId: "session-1",
        sendMessage: jest.fn(),
        isCancelled: jest.fn().mockReturnValue(false),
        taskId: "replicate-task-id",
        ...overrides,
    };
}

//
// Builds a minimal IReplicateDatabaseData for testing.
//
function makeData(overrides: Partial<IReplicateDatabaseData> = {}): IReplicateDatabaseData {
    return {
        sourcePath: "/fake/source",
        destPath: "/fake/dest",
        destEncryptionKey: undefined,
        destS3Key: undefined,
        partial: true,
        force: false,
        ...overrides,
    };
}

describe("replicateDatabaseHandler", () => {
    beforeEach(() => {
        jest.clearAllMocks();

        mockOpenStorage.mockImplementation(async (path: string) => ({
            storage: { __label: `${path}-storage` } as any,
            rawStorage: { __label: `${path}-raw`, write: jest.fn().mockResolvedValue(undefined) } as any,
            encryptionKeyPems: [],
            s3Config: undefined,
            storageOptions: {} as any,
            googleApiKey: undefined,
        }));

        mockReplicate.mockResolvedValue({
            filesImported: 0,
            copiedFiles: 0,
            copiedRecords: 0,
            prunedFiles: [],
        });
    });

    test("opens source storage via openStorage with sourcePath and sourceEncryptionKey", async () => {
        await replicateDatabaseHandler(makeData({ sourceEncryptionKey: "my-source-key" }), makeContext());

        expect(mockOpenStorage).toHaveBeenCalledWith("/fake/source", "my-source-key");
    });

    test("opens destination storage via openStorage with destPath, destEncryptionKey and destS3Key", async () => {
        await replicateDatabaseHandler(makeData({ destEncryptionKey: "dest-key", destS3Key: "dest-s3" }), makeContext());

        expect(mockOpenStorage).toHaveBeenCalledWith("/fake/dest", "dest-key", "dest-s3");
    });

    test("forwards partial flag to replicate() when partial is true", async () => {
        await replicateDatabaseHandler(makeData({ partial: true }), makeContext());

        const options = mockReplicate.mock.calls[0][7];
        expect(options?.partial).toBe(true);
    });

    test("forwards partial flag to replicate() when partial is false", async () => {
        await replicateDatabaseHandler(makeData({ partial: false }), makeContext());

        const options = mockReplicate.mock.calls[0][7];
        expect(options?.partial).toBe(false);
    });

    test("forwards pathFilter to replicate() options", async () => {
        await replicateDatabaseHandler(makeData({ pathFilter: "subdir/" }), makeContext());

        const options = mockReplicate.mock.calls[0][7];
        expect(options?.pathFilter).toBe("subdir/");
    });

    test("emits a replicate-progress task message for each progress callback fired by replicate()", async () => {
        const context = makeContext();

        await replicateDatabaseHandler(makeData(), context);

        const progressCallback = mockReplicate.mock.calls[0][8];
        expect(progressCallback).toBeDefined();
        progressCallback!("Copied 10");
        progressCallback!("Copied 20");

        const sendMessageMock = context.sendMessage as jest.Mock;
        expect(sendMessageMock).toHaveBeenCalledTimes(2);

        const firstMessage = sendMessageMock.mock.calls[0][0] as IReplicateProgressMessage;
        expect(firstMessage).toEqual({
            type: "replicate-progress",
            databasePath: "/fake/source",
            progress: "Copied 10",
        });

        const secondMessage = sendMessageMock.mock.calls[1][0] as IReplicateProgressMessage;
        expect(secondMessage.progress).toBe("Copied 20");
    });

    test("writes encryption.pub to dest raw storage when destination is encrypted", async () => {
        const sourceRawWrite = jest.fn().mockResolvedValue(undefined);
        const destRawWrite = jest.fn().mockResolvedValue(undefined);
        mockOpenStorage
            .mockResolvedValueOnce({
                storage: { __label: "source-storage" } as any,
                rawStorage: { __label: "source-raw", write: sourceRawWrite } as any,
                encryptionKeyPems: [],
                s3Config: undefined,
                storageOptions: {} as any,
                googleApiKey: undefined,
            })
            .mockResolvedValueOnce({
                storage: { __label: "dest-storage" } as any,
                rawStorage: { __label: "dest-raw", write: destRawWrite } as any,
                encryptionKeyPems: [{ privateKeyPem: "priv", publicKeyPem: "pub-from-key" }],
                s3Config: undefined,
                storageOptions: {} as any,
                googleApiKey: undefined,
            });

        await replicateDatabaseHandler(makeData({ destEncryptionKey: "dest-key" }), makeContext());

        expect(destRawWrite).toHaveBeenCalledTimes(1);
        const [path, , buffer] = destRawWrite.mock.calls[0];
        expect(path).toBe(".db/encryption.pub");
        expect(Buffer.isBuffer(buffer)).toBe(true);
        expect((buffer as Buffer).toString("utf-8")).toBe("pub-from-key");
    });

    test("does not write encryption.pub when destination is not encrypted", async () => {
        const destRawWrite = jest.fn().mockResolvedValue(undefined);
        mockOpenStorage.mockImplementation(async (path: string) => ({
            storage: { __label: `${path}-storage` } as any,
            rawStorage: { __label: `${path}-raw`, write: destRawWrite } as any,
            encryptionKeyPems: [],
            s3Config: undefined,
            storageOptions: {} as any,
            googleApiKey: undefined,
        }));

        await replicateDatabaseHandler(makeData({ destEncryptionKey: undefined }), makeContext());

        expect(destRawWrite).not.toHaveBeenCalled();
    });

    test("returns the IReplicationResult from replicate()", async () => {
        mockReplicate.mockResolvedValueOnce({
            filesImported: 5,
            copiedFiles: 3,
            copiedRecords: 2,
            prunedFiles: ["old.jpg"],
        });

        const result = await replicateDatabaseHandler(makeData(), makeContext());

        expect(result).toEqual({
            filesImported: 5,
            copiedFiles: 3,
            copiedRecords: 2,
            prunedFiles: ["old.jpg"],
        });
    });

    test("throws when sourcePath is empty", async () => {
        await expect(replicateDatabaseHandler(makeData({ sourcePath: "" }), makeContext())).rejects.toThrow("sourcePath is required");
    });

    test("throws when destPath is empty", async () => {
        await expect(replicateDatabaseHandler(makeData({ destPath: "" }), makeContext())).rejects.toThrow("destPath is required");
    });
});

File: packages/node-api-zig/src/test/replicate-database.worker.test.zig
const std = @import("std");
const utils = @import("utils-zig");
const node_utils = @import("node-utils-zig");
const task_queue_zig = @import("task-queue-zig");
const node_api = @import("node-api-zig");
const helpers = @import("test-helpers.zig");
const errors = utils.errors;
const replicateDatabaseHandler = node_api.replicate_database_worker.replicateDatabaseHandler;
const TaskContext = task_queue_zig.task_context.TaskContext;
const IReplicateDatabaseData = @import("api-zig").replicate_database_types.IReplicateDatabaseData;

//
// The id of the asset in test/dbs/v6.
//
const ASSET_ID = "89171cd9-a652-4047-b869-1154bf2c95a1";

//
// A task context whose sendMessage records the messages (TypeScript: makeContext with a jest.fn sendMessage).
//
const RecordingContext = struct {
    // Deterministic uuids.
    uuidGenerator: node_utils.test_uuid_generator.TestUuidGenerator,

    // Deterministic time.
    timestampProvider: node_utils.test_timestamp_provider.TestTimestampProvider,

    // The task context.
    context: TaskContext,

    // The recorded messages, as JSON text.
    messages: std.ArrayList([]const u8),

    // Allocates the recorded messages.
    allocator: std.mem.Allocator,

    //
    // Records a message.
    //
    fn sendMessage(context: ?*anyopaque, message: std.json.Value) void {
        const self: *RecordingContext = @ptrCast(@alignCast(context.?));
        const text = std.json.Stringify.valueAlloc(self.allocator, message, .{}) catch {
            return;
        };
        self.messages.append(self.allocator, text) catch {};
    }
};

//
// Builds a minimal ITaskContext for testing.
//
fn makeContext(allocator: std.mem.Allocator, io: std.Io) !*RecordingContext {
    _ = try helpers.setupEnvironment(io);
    const recording = try allocator.create(RecordingContext);
    recording.* = .{
        .uuidGenerator = try node_utils.test_uuid_generator.TestUuidGenerator.init(allocator),
        .timestampProvider = .{},
        .context = undefined,
        .messages = .empty,
        .allocator = allocator,
    };
    recording.context = TaskContext.init(recording.uuidGenerator.uuidGenerator(), recording.timestampProvider.timestampProvider(), "session-1", "replicate-task-id", .{ .context = recording, .function = RecordingContext.sendMessage });
    return recording;
}

//
// The directories of a test: a copy of test/dbs/v6 as the source and a path for the destination.
//
const Directories = struct {
    // The directory holding everything.
    root: []const u8,

    // The source database.
    source: []const u8,

    // The destination database (does not exist yet).
    dest: []const u8,
};

//
// Creates the directories of a test.
//
fn makeDirectories(allocator: std.mem.Allocator, io: std.Io) !Directories {
    const sourceDir = try helpers.copyTestDatabase(allocator, io, "v6");
    const root = std.fs.path.dirname(sourceDir).?;
    return .{ .root = root, .source = sourceDir, .dest = try std.fmt.allocPrint(allocator, "{s}/dest", .{root}) };
}

//
// Builds a minimal IReplicateDatabaseData for testing, as the JSON the queue carries.
//
fn makeData(allocator: std.mem.Allocator, data: IReplicateDatabaseData) !std.json.Value {
    return node_api.replicate_database.replicateDatabaseDataToJson(allocator, data);
}

//
// Reads a file of a directory.
//
fn readFileIn(allocator: std.mem.Allocator, io: std.Io, dir: []const u8, name: []const u8) ![]const u8 {
    return helpers.readFile(allocator, io, try std.fmt.allocPrint(allocator, "{s}/{s}", .{ dir, name }));
}

test "opens source storage via openStorage with sourcePath and sourceEncryptionKey" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);
    const keyFile = helpers.KEYS_DIR ++ "/ts-private.pem";

    // Make an encrypted copy, then replicate the encrypted copy back to a plain database with the source key.
    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .destEncryptionKey = keyFile, .partial = false, .force = false }), recording.context.taskContext());
    const plainDir = try std.fmt.allocPrint(allocator, "{s}/plain", .{dirs.root});
    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.dest, .destPath = plainDir, .sourceEncryptionKey = keyFile, .partial = false, .force = false }), recording.context.taskContext());

    const assetPath = "asset/" ++ ASSET_ID;
    try std.testing.expectEqualStrings(try readFileIn(allocator, io, dirs.source, assetPath), try readFileIn(allocator, io, plainDir, assetPath));
    try std.testing.expect(!std.mem.eql(u8, try readFileIn(allocator, io, dirs.source, assetPath), try readFileIn(allocator, io, dirs.dest, assetPath)));
}

test "opens destination storage via openStorage with destPath, destEncryptionKey and destS3Key" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);
    const keyFile = helpers.KEYS_DIR ++ "/ts2-private.pem";

    // destS3Key is only used for s3: destinations (see resolve-storage-credentials tests); here it is ignored.
    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .destEncryptionKey = keyFile, .destS3Key = "dest-s3", .partial = false, .force = false }), recording.context.taskContext());

    const dest = try node_api.open_storage.openStorage(allocator, io, dirs.dest, keyFile, null);
    const assetPath = "asset/" ++ ASSET_ID;
    try std.testing.expectEqualStrings(try readFileIn(allocator, io, dirs.source, assetPath), (try dest.storage.read(allocator, io, assetPath)).?);
}

test "forwards partial flag to replicate() when partial is true" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);

    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .partial = true, .force = false }), recording.context.taskContext());

    const dest = try node_api.open_storage.openStorage(allocator, io, dirs.dest, null, null);
    const destTree = (try node_api.tree.loadMerkleTree(allocator, io, dest.storage)).?;
    try std.testing.expect(node_api.media_file_database.isPartialDatabase(destTree.databaseMetadata));
    try std.testing.expect(!helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/asset/{s}", .{ dirs.dest, ASSET_ID })));
    try std.testing.expect(helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/.db/bson/collections/metadata/shards/96.dat", .{dirs.dest})));
    try std.testing.expect(!helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/.db/bson/collections/metadata/shards/96", .{dirs.dest})));
}

test "forwards partial flag to replicate() when partial is false" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);

    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .partial = false, .force = false }), recording.context.taskContext());

    const dest = try node_api.open_storage.openStorage(allocator, io, dirs.dest, null, null);
    const destTree = (try node_api.tree.loadMerkleTree(allocator, io, dest.storage)).?;
    try std.testing.expect(!node_api.media_file_database.isPartialDatabase(destTree.databaseMetadata));
    try std.testing.expect(helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/asset/{s}", .{ dirs.dest, ASSET_ID })));
    try std.testing.expect(helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/.db/bson/collections/metadata/shards/96", .{dirs.dest})));
}

test "forwards pathFilter to replicate() options" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);

    const output = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .partial = false, .force = false, .pathFilter = "display/" ++ ASSET_ID }), recording.context.taskContext());

    try std.testing.expectEqual(@as(i64, 1), output.object.get("copiedFiles").?.integer);
    try std.testing.expect(helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/display/{s}", .{ dirs.dest, ASSET_ID })));
    try std.testing.expect(!helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/thumb/{s}", .{ dirs.dest, ASSET_ID })));
}

test "emits a replicate-progress task message for each progress callback fired by replicate()" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);

    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .partial = false, .force = false }), recording.context.taskContext());

    try std.testing.expectEqual(@as(usize, 4), recording.messages.items.len);
    const firstMessage = try std.fmt.allocPrint(allocator, "{{\"type\":\"replicate-progress\",\"databasePath\":\"{s}\",\"progress\":\"Copied 1\"}}", .{dirs.source});
    try std.testing.expectEqualStrings(firstMessage, recording.messages.items[0]);
    const lastMessage = try std.fmt.allocPrint(allocator, "{{\"type\":\"replicate-progress\",\"databasePath\":\"{s}\",\"progress\":\"Copied 3 files, 1 records\"}}", .{dirs.source});
    try std.testing.expectEqualStrings(lastMessage, recording.messages.items[3]);
}

test "writes encryption.pub to dest raw storage when destination is encrypted" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);

    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .destEncryptionKey = helpers.KEYS_DIR ++ "/ts-private.pem", .partial = true, .force = false }), recording.context.taskContext());

    try std.testing.expectEqualStrings(
        try helpers.readFile(allocator, io, helpers.KEYS_DIR ++ "/ts-public.pem"),
        try readFileIn(allocator, io, dirs.dest, ".db/encryption.pub"),
    );
}

test "does not write encryption.pub when destination is not encrypted" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);

    _ = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .partial = true, .force = false }), recording.context.taskContext());

    try std.testing.expect(!helpers.fileExists(io, try std.fmt.allocPrint(allocator, "{s}/.db/encryption.pub", .{dirs.dest})));
}

test "returns the IReplicationResult from replicate()" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    const dirs = try makeDirectories(allocator, io);
    defer helpers.removeTempDir(io, dirs.root);

    const output = try replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = dirs.source, .destPath = dirs.dest, .partial = false, .force = false }), recording.context.taskContext());

    try std.testing.expectEqualStrings("{\"filesImported\":1,\"copiedFiles\":3,\"copiedRecords\":1,\"prunedFiles\":[]}", try std.json.Stringify.valueAlloc(allocator, output, .{}));
}

test "throws when sourcePath is empty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    try std.testing.expectError(error.Thrown, replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = "", .destPath = "/fake/dest", .partial = true, .force = false }), recording.context.taskContext()));
    try std.testing.expectEqualStrings("sourcePath is required", errors.lastErrorMessage());
}

test "throws when destPath is empty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const io = std.testing.io;
    const recording = try makeContext(allocator, io);
    try std.testing.expectError(error.Thrown, replicateDatabaseHandler(allocator, io, try makeData(allocator, .{ .sourcePath = "/fake/source", .destPath = "", .partial = true, .force = false }), recording.context.taskContext()));
    try std.testing.expectEqualStrings("destPath is required", errors.lastErrorMessage());
}
